Standard text components

plfiorini opened the pull request:
Add a bunch of standard components for text.

Relying on FluidStyle properties for standard font settings
might be confusing for new users while a set of components will
make the choice obvious.

Individual text components allows us to set font properties
and line heights according to the guidelines automatically.

plfiorini commented:
This is how it looks like:

![fluidtypography](https://cloud.githubusercontent.com/assets/843710/16893845/ae7ed864-4b44-11e6-9c46-43f5990dac8b.png)

If you like this PR I can go further and a C++ template with the level property

plfiorini commented:
Probably we should rename all components with the Label suffix, like DisplayLabel, HeadlineLabel, SubheadingLabel, ....

iBeliever commented:
I like the idea, especially if you rename everything to add a Label suffix. However, the one concern I have is upstreaming font styles to Controls 2.

There's [QTBUG-51906](https://bugreports.qt.io/browse/QTBUG-51906) filed against the Material theme for font styles, so it would be nice to implement the styles in a way that can be upstreamed easily. I implemented the font style properties the way I did because I was thinking they could converted to C++ properties on the `Material` attached style object, and used like `font: Material.display1Font`, just as you now do `color: Material.primaryTextColor`. 

What about continuing to use the FluidStyle properties with the hope that we'd upstream them, but then also provide our `TitleLabel`, `DisplayLabel`, `CaptionLabel`, etc. that uses those font styles plus setting any additional properties like `lineHeight`?

iBeliever commented on ui/+material/DisplayText.qml:
```diff
@@ -0,0 +1,42 @@
+import Fluid.Core 1.0
+
+T.Label {
+    property int level: 1
```
You could add `onLevelChanged` and do `console.error` if level < 1 or > 4.

plfiorini replied:
DONE

iBeliever commented on ui/+material/Subheading.qml:
```diff
@@ -0,0 +1,23 @@
+import Fluid.Core 1.0
+
+T.Label {
+    property int level: 1
```
As far as I know, there's only Subheading font style, and no level variations of it.

plfiorini replied:
It's mentioned here: https://material.google.com/style/typography.html#typography-line-height but not in early paragraphs.
What should I do?

iBeliever replied:
Interesting, I was looking at a section higher up were that wasn't mentioned. Let's leave it then. My guess is it's not mentioned in the first section because that section only applies to font size, and the only difference between 1 and 2 is the leading value.

plfiorini replied:
OK

iBeliever commented on ui/BodyText.qml:
```diff
@@ -0,0 +1,29 @@
+    \endcode
+*/
+T.Label {
+    property int level: 1
```
This isn't be used, but should. There are body 1 and body 2 fonts, with body 2 being medium weight.

iBeliever commented on ui/BodyText.qml:
```diff
@@ -0,0 +1,29 @@
+    }
+    \endcode
+*/
+T.Label {
```
Missing link color?

iBeliever commented on ui/Caption.qml:
```diff
@@ -0,0 +1,27 @@
+    \endcode
+*/
+T.Label {
+    font.pixelSize: 10
```
Should be 12

iBeliever commented on ui/Caption.qml:
```diff
@@ -0,0 +1,27 @@
+    }
+    \endcode
+*/
+T.Label {
```
Missing linkColor

iBeliever commented:
Do you need `color: Material.primaryTextColor` in all of the +material labels?

plfiorini commented:
QQC2 label for material doesn't set the color, I would not set here either.

plfiorini commented:
The Qt bug has a code review linked and that seems to implement additional font styles like small and large variations, not quite like what we want to achieve here.

Regardless we can keep FluidStyle, this PR is not meant to completely replace it, the goal is to add convenient components only.

plfiorini commented:
PR update:

* Renamed all components with Label suffix
* Added a unit test for Material guidelines, though I cannot use +material from tests so it will always pass
* Added errors for invalid levels

I don't see the need for a C++ label template now this means that we use the original QQC2 template and add a level property.

iBeliever commented:
Merged
